src/main/scala/org/constellation/domain/healthcheck/ReconciliationRound.scala

```diff
@@ -0,0 +1,244 @@
+import org.constellation.schema.NodeState.{canActAsJoiningSource, isInitiallyJoining, isInvalidForJoining}
+import org.constellation.schema.consensus.RoundId
+import org.constellation.schema.{Id, NodeState}
+
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+
+class ReconciliationRound[F[_]](
+  ownId: Id,
+  cluster: Cluster[F],
+  val startedAtSecond: Long
+)(
+  apiClient: ClientInterpreter[F],
+  unboundedBlocker: Blocker
+)(implicit F: Concurrent[F], CS: ContextShift[F], C: Clock[F], T: Timer[F]) {
+  import ReconciliationRound._
+
+  val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+
+  val retryAttempts = 2
+  val retryAfter: FiniteDuration = 60.seconds
+
+  private[healthcheck] val reconciliationResult: Ref[F, Option[ReconciliationResult]] =
+    Ref.unsafe(None)
+
+//  private[healthcheck] val firstClusterAlignmentResults: Ref[F, Option[Map[Id, List[ClusterAlignmentResult]]]] =
+//    Ref.unsafe(None)
+
+//  private[healthcheck] val lastClusterAlignmentResults: Ref[F, Option[Map[Id, List[ClusterAlignmentResult]]]] =
+//    Ref.unsafe(None)
+
+  def getReconciliationResult(): F[Option[ReconciliationResult]] = reconciliationResult.get
+
+  def start(): F[Fiber[F, Unit]] = F.start(run())
+
+  private def run(
+    attemptsLeft: Int = retryAttempts,
+    firstClusterAlignmentResult: Option[Map[Id, List[ClusterAlignmentResult]]] = None,
+    lastClusterAlignmentResult: Option[Map[Id, List[ClusterAlignmentResult]]] = None
+  ): F[Unit] =
+    for {
+      isFirstRun <- F.delay(attemptsLeft == retryAttempts)
+      isLastRun <- F.delay(attemptsLeft == 0)
+      _ <- if (isFirstRun) F.unit else T.sleep(retryAfter)
+      result <- runReconciliation()
+      _ <- result match {
+        case Left(unresponsivePeers) =>
+          reconciliationResult.modify(
+            // shouldn't we only run healthcheck for nodes that didn't respond instead of running consensus right away?
+            // in consequence I think we should also try to calculate cluster state from the responses we got and start consensuses for these
+            _ => (ReconciliationResult(unresponsivePeers, Set.empty, lastClusterAlignmentResult).some, ())
+          )
+
+        case Right(clusterAlignmentResults) =>
+          if (areAllNodesAligned(clusterAlignmentResults))
+            reconciliationResult.modify(
+              _ => (ReconciliationResult(Set.empty, Set.empty, clusterAlignmentResults.some).some, ())
+            )
+          else if (!isLastRun)
+            run(
+              attemptsLeft - 1,
+              if (isFirstRun) clusterAlignmentResults.some else firstClusterAlignmentResult,
+              clusterAlignmentResults.some
+            )
+          else {
+            for {
+              initiallyMisaligned <- F.delay(
+                firstClusterAlignmentResult.map(fetchMisalignedPeers).getOrElse(Set.empty)
+              )
+              currentlyMisaligned <- F.delay(fetchMisalignedPeers(clusterAlignmentResults))
```

Comment:
why is it considered a side effect and is delayed in a monad? imo these are pure functions so there is no need for delaying

Reply:
The same as above. Had a situation that a function that seemed pure didn't execute unless it was suspended in F.delay. Maybe I missed something back then and it was a different issue, or impure afterall.

Reply:
@kpudlik regarding wrapping in pure